shark-hprof/src/main/java/shark/HprofDeobfuscator.kt

```diff
@@ -0,0 +1,237 @@
+                  is ObjectArrayDumpRecord -> maxId.coerceAtLeast(record.id)
+                  is ObjectArraySkipContentRecord -> maxId.coerceAtLeast(record.id)
+                  is PrimitiveArrayDumpRecord -> maxId.coerceAtLeast(record.id)
+                  is PrimitiveArraySkipContentRecord -> maxId.coerceAtLeast(record.id)
+                }
+              }
+            }
+          })
+      }
+
+    return maxId
+  }
+
+  @Suppress("LongParameterList")
+  private fun writeHprofRecords(
+    inputHprofFile: File,
+    outputHprofFile: File,
+    proguardMapping: ProguardMapping,
+    hprofStringCache: MutableMap<Long, String>,
+    classNames: MutableMap<Long, Long>,
+    firstId: Long
+  ): File {
+    var id = firstId
+
+    Hprof.open(inputHprofFile)
+      .use { hprof ->
+        val reader = hprof.reader
+        HprofWriter.open(
+          outputHprofFile,
+          identifierByteSize = reader.identifierByteSize,
+          hprofVersion = hprof.hprofVersion
+        ).use { writer ->
+          reader.readHprofRecords(setOf(HprofRecord::class),
+            OnHprofRecordListener { _,
+              record ->
+              // HprofWriter automatically emits HeapDumpEndRecord, because it flushes
+              // all continuous heap dump sub records as one heap dump record.
+              if (record is HeapDumpEndRecord) {
+                return@OnHprofRecordListener
+              }
+
+              when (record) {
+                is StringRecord -> {
+                  writer.write(
+                    createDeobfuscatedStringRecord(record, proguardMapping, hprofStringCache)
+                  )
+                }
+                is HeapDumpRecord -> {
+                  when (record) {
+                    is ObjectRecord -> {
+                      when (record) {
+                        is ClassDumpRecord -> {
+                          val recordsToWrite = createDeobfuscatedClassDumpRecord(
+                            record, proguardMapping, hprofStringCache, classNames, id
+                          )
+                          id += recordsToWrite.size
+                          recordsToWrite.forEach {
+                            writer.write(it)
+                          }
+                        }
+                        else -> writer.write(record)
+                      }
+                    }
+                    else -> writer.write(record)
+                  }
+                }
+                else -> writer.write(record)
+              }
+            })
+            }
+      }
+
+    return outputHprofFile
+  }
+
+  private fun createDeobfuscatedStringRecord(
+    record: StringRecord,
+    proguardMapping: ProguardMapping,
+    hprofStringCache: MutableMap<Long, String>
+  ): StringRecord {
+    val obfuscatedName = hprofStringCache[record.id]!!
+    return StringRecord(
+      record.id, proguardMapping.deobfuscateClassName(obfuscatedName)
```

Comment:
I'm surprised this works as it only accounts for class names. What about field names?

Reply:
I think I get it: this is relying on the fact that ProguardMapping will return `obfuscatedName` if no mapping is found.

What if there's a field named `a` and a class named `a`? Does proguard always add a package prefix?

Reply:
In order to deobfuscate a field, we need an obfuscated field name + obfuscated class name. If we have only a field name then we can't deobfuscate it.
And the rest is correct, it greedily tries to deobfuscate each record hoping that it's a class name and if it's not present in the mapping then the original name will be returned.

Do you think that it would be better to explicitly handle class names?

Reply:
> In order to deobfuscate a field, we need an obfuscated field name + obfuscated class name. If we have only a field name then we can't deobfuscate it.

Right, but `StringRecord` does make any distinction AFAIK.

E.g. if you create a class named `a` (no package name) and a field named `a` in it (ie `a.a`), then I expect to see only one StringRecord `a`, referenced by both LoadClass and the fields in ClassDump.

Reply:
I think it'll still work correctly. During write phase we're reading from an original, obfuscated hprof and we have a map filled with record.id->string entries.
Now if we encounter a class named `a` then we'll deobfuscate it to let's say `Foo` and write such record into a new hprof file. If we encounter a `ClassDump` with a field named `a` (the same StringRecord `a` as before) then we look into our map and see obfuscated class name and field name for this field. So we can deobfuscate it and create a new `StringRecord` for this field name.

Reply:
ha, nice. But this might break if we stop creating new records when the fields can't be deobfuscated (or, rather, we'd have to check that this string record has been kept in the new hprof)

Reply:
So you're saying that we may  end up with deobfuscating a class name from `a` to `Foo` and then a field `a` will be also printed as `Foo` in the deobfuscated heap dump, correct? So we either  have to generate a new StringRecord for each field like we did before or check if the StringRecord with given id was already deobfuscated (as a ClassName) and in this case generate a new StringRecord for a field so that it doesn't collide. If that's correct then which apporach do you prefer? First one is simpler but may end up making the resulting hprof bigger while the  second one will be more complicated but may save some space. Hard to say how much it'll affect the size though.

Reply:
I think maybe going back to duplicating string records is the simplest / best here. The hprof size shouldn't be too much of a concern, and if it is we could always add a pass that removes unused records.

Reply:
Reverted the changes and now we're always duplicating the records.